2D power spectrum structures

damonge opened the pull request:
Addresses part of #511 , associated with the generalized power spectrum structures

I'd suggest deferring the merging of this until all PRs related to the CCL paper have been merged.

At the pyhton level, the generalized structures are defined and described in [this file](https://github.com/LSSTDESC/CCL/blob/power_spectrum_structs/pyccl/p2d.py) (look at the docstrings for the time being).

The interface of these objects with the Limber integrator is described [here](https://github.com/LSSTDESC/CCL/blob/2e0c1f02f30a86a77ca4abb364293674f1b43f0d/pyccl/cls.py#L297).

tmcclintock commented:
Looked at the docstring. This would be a good first docstring to note whether distances are comoving or not. For instance, here, the wavenumbers are comoving, so at [this line](https://github.com/LSSTDESC/CCL/blob/50b02858d284b049c81eef0d31b889f940d47f97/pyccl/p2d.py#L25) we could write `(in units of Mpc^-1; comoving)` instead of just `(in units of Mpc^-1)`.

damonge commented:
@tmcclintock you'd have to be kind of evil to ask for not-comoving k, but granted, we can document that better. I wonder if there's 100 other places where the same thing happens.

tmcclintock commented:
@damonge of course :), but I think back to a younger me who wondered at one time "does CLASS return k in comoving or physical coordinates?" and spent an hour finding the answer. Clarity never hurt, IMO.

damonge commented:
No, definitely. Will do!

elikrause commented on src/ccl_p2d.c:
```diff
@@ -0,0 +1,182 @@
+  }
+  else if(is_lok) {
+    pk_post=pk_pre;
+    pk_post=pk_pre;
```
(duplicate line)

damonge replied:
Well spotted

elikrause commented on src/ccl_p2d.c:
```diff
@@ -0,0 +1,182 @@
+	return -1;
+      }
+      pk_post+=pd*dlk;
+      if(psp->extrap_order_hik>1) {
```
I should know this - is extrap_order_* restricted to {0,1,2}?

damonge replied:
Correct. This was documented in the C docstrings, and I've now added documentation in the python functions too.

damonge commented on src/ccl_cls.c:
```diff
@@ -887,27 +887,39 @@ static void get_k_interval(ccl_cosmology *cosmo,CCL_ClWorkspace *w,
+  if(psp==NULL) {
+    if (!cosmo->computed_power) ccl_cosmology_compute_power(cosmo, status);
+    // Return if compilation failed
+    if (!cosmo->computed_power) return NAN;
```
It is handled by the python status check. However, to make the process faster I've changed things a bit so that, the first time this happens, it goes straight back to python without trying to compute anything else.

damonge commented:
@elikrause thanks a lot for your review!

I've added some extra documentation and warnings about power spectrum sampling (in the python docstrings).

About ccl_sigmaR etc.: now CCL stores the linear and non-linear power spectra as two internal `ccl_p2d_t` structures, which are evaluated whenever the functions `ccl_linear_matter_power` or `ccl_nonlin_matter_power` are called, so those functions will still work as before. I have not added functionality to modify the internal matter power spectrum structures manually from python, but I don't know if this is something we wanted to do (i.e. the main use of the Pk2D objects now is to pass them to the Limber integrator, but not to mass functions etc.).

elikrause commented:
> About ccl_sigmaR etc.: now CCL stores the linear and non-linear power spectra as two internal `ccl_p2d_t` structures, which are evaluated whenever the functions `ccl_linear_matter_power` or `ccl_nonlin_matter_power` are called, so those functions will still work as before. 

Yes, I agree that they should still work as before as it simply calls the z=0 power spectrum, and factoring in the growth. I believe this assumption in current variance calculation + halo model routines could lead to inconsistencies if a 2D power spectrum is supplied by the user, but then only used at z=0 and rescaled by growth factor in the halo calculations.

damonge commented:
@elikrause thanks a lot!

The user is not able to alter the internal linear and non-linear matter power spectra. They are just able to create new Pk2D objects and to pass them to the limber integrator. So I think this shouldn't lead to inconsistencies.

Now that you mention it, however, the fact that we extrapolate with linear growth the z=0 power spectrum for halo quantities is kind of problematic for e.g. cosmologies with neutrinos, so that may be something we should fix in a different PR...

elikrause commented:
@damonge 
> The user is not able to alter the internal linear and non-linear matter power spectra. They are just able to create new Pk2D objects and to pass them to the limber integrator. So I think this shouldn't lead to inconsistencies.

Thanks, it seems to me that we agree on the workings of CCL, just call it differently. It might be useful to comment somewhere in the description that Pk2D is *only* used by the 2-pt Limber integrator, not in the rest of the code. At least I briefly had the idea that Pk2D might be a method to do variance/halo model computations where scale and redshift dependence don't factorize. I agree now that this should be a separate PR.

I approve, but as this is a substantial PR, a second review might be useful.

beckermr commented on include/ccl_emu17.h:
```diff
@@ -11,11 +11,12 @@ CCL_BEGIN_DECLS
  * Emulator power spectrum
  * Obtain P(k,z) [Mpc^3] for a given set of input parameters.
  * @param xstarin vector of input parameters for the emulator, including redshift. 
+ * @param nkemu number of elements to which Pkemu has been allocated.
```
What does this mean?

damonge replied:
It's the size of the next argument (`Pkemu`)

beckermr replied:
Ahhh. The doc string confused me.

beckermr commented on include/ccl_params.h:
```diff
@@ -44,6 +44,11 @@ typedef struct ccl_spline_params {
+int ccl_get_pk_spline_na(void);
+int ccl_get_pk_spline_nk(void);
+void ccl_get_pk_spline_a_array(int ndout,double* doutput,int *status);
+void ccl_get_pk_spline_lk_array(int ndout,double* doutput,int *status);
```
Why are these prototypes here and not in the 2D power spectrum header file?

damonge replied:
These functions are basically a patch to grab the internal CCL spline parameters and make them visible to python (essentially this gets superseded by your PR, I think). Because they just access the contents of `ccl_spline_params`, I think it's more appropriate for them to be in ccl_params.h

beckermr replied:
Got it. In my PR to make the spline parameters changeable, I moved these structs to `ccl_core.h` and deleted the `ccl_params.h` file. Could we move these prototypes and the function defs to `ccl_core.h` and `ccl_core.c` to avoid the merge conflicts?

damonge replied:
Done in my last commit

beckermr replied:
These prototypes are still here. Let's delete the params.h file and just use core.h.

beckermr commented on src/ccl_power.c:
```diff
@@ -406,194 +443,125 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
+    if (parser_free(&fc)== _FAILURE_) {
+      *status = CCL_ERROR_CLASS;
+      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error freeing CLASS parser\n");
+    }
```
I don't see the place where the parser is freed if status is non-zero. Did we reintroduce this memory leak?

damonge replied:
Right, the whole file has been refactored (that's why I didn't push to solving this earlier, since I thought this would be fixed when this got merged). This is now done in https://github.com/LSSTDESC/CCL/blob/3a3f5e9a2194465312e03b7782134ea0876eb07c/src/ccl_power.c#L446

beckermr commented on pyccl/cls.py:
```diff
@@ -316,18 +328,18 @@ def angular_cl(cosmo, cltracer1, cltracer2, ell,
-            cosmo, clt1, clt2, l_limber, l_logstep,
+            cosmo, clt1, clt2, psp, l_limber, l_logstep,
             l_linstep, ell, ell.size, status)
     else:
```
While you are here, using `np.atleast_1d` will cover all three of these cases in this if else block

damonge replied:
Done in last commit. Note that there's a lot of these in most of the python functions.

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,144 @@
+    """
+    def __init__(self, pkfunc=None, a_arr=None, lk_arr=None, pk_arr=None,
+                 is_logp=True, interp_order_lok=1, interp_order_hik=2):
+        """Constructor for Pk2D objects.
```
Doc strings for the init go in the class level doc string, not on the init.

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,144 @@
+        if(pkfunc is None):  # Initialize power spectrum from 2D array
+            # Make sure input makes sense
+            if (a_arr is None) or (lk_arr is None) or (pk_arr is None):
+                raise TypeError("If you do not provide a function, "
```
This should be a ValueError.

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,144 @@
+            try:
+                pkfunc(k=np.array([1E-2, 2E-2]), a=0.5)
+            except Exception:
+                raise TypeError("Can't use input function")
```
This is also a ValueError.

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,144 @@
+        else:
+            raise NotImplementedError("Currently we need a cosmology to "
+                                      "extrapolate growth")
+            cospass = None
```
This block means you have to pass a cosmology even though the parameter argument is optional.

damonge replied:
Made non-optional

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,144 @@
+        check(status, cosmo)
+
+        return f
+        raise NotImplementedError("Not implemented yet")
```
Remove this line here.

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,144 @@
+        elif isinstance(k, np.ndarray):
+            f, status = lib.p2d_eval_multi(self.psp, np.log(k), a, cospass,
+                                           k.size, status)
+        else:
```
again `np.atleast_1d` might cover some of these cases.

damonge replied:
Fixed in last commit

beckermr commented on src/ccl_emu17.c:
```diff
@@ -138,21 +138,30 @@ static void emuInit() {
-void ccl_pkemu(double *xstar, double **ystar, int* status, ccl_cosmology* cosmo) {
-    
+void ccl_pkemu(double *xstar, int sizeofystar, double *ystar, int* status, ccl_cosmology* cosmo)
+{
```
Why the interface change here?

damonge replied:
Before this routine set `*ystar` to `NULL` to mark when it failed, which could create memory leaks. This is now handled through the `status` flag, and I pass the size of `ystar` to add an extra check that it has been correctly allocated.

beckermr commented on src/ccl_p2d.c:
```diff
@@ -0,0 +1,181 @@
+  else if(is_hiz) { //Are we below the interpolation range in a?
+    if(psp->extrap_linear_growth==ccl_p2d_no_extrapol) {
+      *status=CCL_ERROR_SPLINE_EV;
+      return NAN;
```
Why does one statement return -1 and the other NaN?

damonge replied:
Homogenized to NAN

beckermr commented on src/ccl_p2d.c:
```diff
@@ -0,0 +1,181 @@
+  int is_loz= a>psp->amax;
+
+  if(is_loz) { //Are we above the interpolation range in a?
+    *status=CCL_ERROR_SPLINE_EV;
```
Do you not need an extrapolation check here to like the one below

damonge replied:
We do not have a method to extrapolate below the maximum redshift of the spline. The logic is that it may be safe to assume linear growth at high-redshift, but not at low redshift. So in this case we just return with an error.

beckermr replied:
got it.

beckermr commented on src/ccl_p2d.c:
```diff
@@ -0,0 +1,181 @@
+  int spstatus=gsl_spline2d_eval_e(psp->pk,lk_ev,a_ev,NULL,NULL,&pk_pre);
+  if(spstatus) {
+    *status=CCL_ERROR_SPLINE_EV;
+    return -1;
```
Is this -1 or NaN?

damonge replied:
Homogenized to NAN

beckermr commented on src/ccl_params.c:
```diff
@@ -0,0 +1,66 @@
+      doutput[ii]=log(d[ii]);
+  }
+  free(d);
+}
```
Let's move these functions to `ccl_core.c`

damonge replied:
Done in my last commit

beckermr commented on tests/ccl_test_correlation_3d.c:
```diff
@@ -83,9 +83,9 @@ static void compare_correlation_3d(int i_model,struct corrs_3d_data * data)
+  double CORR_TOLERANCE1[6] = {4.0e-2, 1.0e-2, 5.0e-3, 4.0e-3, 1.0e-3, 8.0e-4};
   // tolerence on abs difference in r^2 xi(r) for the range r = 50 - 250 Mpc (100 points in r) for z=0,1,2,3,4,5
-  double CORR_TOLERANCE2[6] = {3.15e-2, 9.87e-3, 5.04e-3, 3.05e-3, 9.98e-4, 7.14e-4};
+  double CORR_TOLERANCE2[6] = {4.0e-2, 1.0e-2, 5.0e-3, 4.0e-3, 1.0e-3, 8.0e-4};
```
Should we be concerned about these tolerances going up?

damonge replied:
These were set originally to **exactly** the tolerances that were achieved at the time. When the P(k) splines change, they move up and down slightly, so I've just made them all larger to allow future changes in the P(k) part from affecting this test. 
We're trying to test the P(k) -> xi(r) code here, so this test should be robust against changes in the P(k) calculation.

beckermr replied:
Right. This test is pretty useless TBH.

beckermr commented on tests/ccl_test_p2d.c:
```diff
@@ -0,0 +1,172 @@
+  ASSERT_DBL_NEAR(1,pk/pk_model_analytical(exp(lktest),atest));
+
+  //Evaluate at very low z and see if it checks out
+  double alo=0.02;
```
Lowz is a ~ 1.

damonge replied:
Yes, sorry, this was meant to read high z

beckermr replied:
cool. Let's change the comment in the code.

beckermr commented on tests/ccl_test_p2d.c:
```diff
@@ -0,0 +1,172 @@
+  ASSERT_DBL_NEAR(1,pk/pk_model_analytical(exp(lkhi),atest));
+
+  //Evaluate at very high k and see if it checks out
+  double lklo=data->lk_arr[0]*1.1;
```
Do you need to divide by 1.1 here?

damonge replied:
Correct! Fixed

beckermr commented on tests/ccl_test_p2d.py:
```diff
@@ -0,0 +1,108 @@
+def check_p2d_init():
+    """
+    Check that background and growth functions can be run.
+    """
```
Why is this function empty?

damonge replied:
Sorry, some residuals from copy-paste. Removed

beckermr commented on src/ccl_power.c:
```diff
@@ -406,194 +443,123 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
+  double *x=NULL, *z=NULL, *y2d_lin=NULL, *y2d_nl=NULL;
+  if (*status == 0) {
+    //CLASS calculations done - now allocate CCL splines
+    kmin = 2*exp(sp.ln_k[0]);
```
Why the factor of 2 here?

beckermr replied:
Or should the min be the CCL value?

damonge replied:
I'd point you to Elisabeth or Elisa (they implemented the CLASS interface, and this particular line hasn't been changed in this PR). I imagine this is to make sure that we don't try to evaluate things below the minimum k used by CLASS.

beckermr replied:
@elisachisari @elikrause can you answer the question here?

beckermr commented on src/ccl_power.c:
```diff
@@ -406,194 +443,123 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
+  if (*status == 0) {
+    //CLASS calculations done - now allocate CCL splines
+    kmin = 2*exp(sp.ln_k[0]);
+    kmax = ccl_splines->K_MAX_SPLINE;
```
Why are we not pulling kmax from the CLASS spectra?

damonge replied:
Again, Elisabeth and Elisa are the right people to ask here.

damonge replied:
Done in last commit

beckermr commented on src/ccl_power.c:
```diff
@@ -406,194 +443,123 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
+      *status = CCL_ERROR_CLASS;
+      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): K_MIN is less than CLASS's kmin. Not yet supported for nonlinear P(k).\n");
+    }
   }
```
Why are testing this if we set it above?

damonge replied:
No idea. This is a leftover from the previous code. Removed in my last commit

beckermr commented on src/ccl_power.c:
```diff
@@ -873,120 +836,98 @@ static void ccl_cosmology_compute_power_eh(ccl_cosmology * cosmo, int * status)
+  
+  if(*status==0) {
+    if(cosmo->config.baryons_power_spectrum_method==ccl_bcm)
+      correct_bcm(cosmo,na,z,nk,x,y2d,status);
```
are we applying BCM to the linear power spectrum here?

damonge replied:
We only apply it to whatever array is gonna end up in the non-linear power spectrum spline

beckermr commented on src/ccl_power.c:
```diff
@@ -1037,130 +976,100 @@ static void ccl_cosmology_compute_power_bbks(ccl_cosmology * cosmo, int * status
+  
+  if(*status==0) {
+    if(cosmo->config.baryons_power_spectrum_method==ccl_bcm)
+      correct_bcm(cosmo,na,z,nk,x,y2d,status);
```
Again we should not be applying the BCM to the linear P(k).

damonge replied:
Agreed :-)

beckermr replied:
Wow ok. So actually if you specify EH98 for the transfer function, you get that for the non-linear and linear power?

beckermr replied:
@damonge pinging you to make sure you see this.

beckermr replied:
discussed below.

damonge replied:
Sorry, didn't see this comment:

>So actually if you specify EH98 for the transfer function, you get that for the non-linear and linear power?

Yes, that's correct

beckermr commented on src/ccl_power.c:
```diff
@@ -1238,116 +1135,150 @@ static void ccl_cosmology_compute_power_emu(ccl_cosmology * cosmo, int * status)
+  // Prepare to run CLASS for linear scales
+  ccl_fill_class_parameters(cosmo,&fc,parser_length, status);
+
+  if (*status != CCL_ERROR_CLASS)
```
Random question. Does the emulator not have a linear prediction to go with it?

damonge replied:
Nope, we use CLASS

beckermr commented on src/ccl_power.c:
```diff
@@ -1238,116 +1135,150 @@ static void ccl_cosmology_compute_power_emu(ccl_cosmology * cosmo, int * status)
+  if (*status == 0) {
+    //CLASS calculations done - now allocate CCL splines
+    kmin = 2*exp(sp.ln_k[0]);
+    kmax = ccl_splines->K_MAX_SPLINE;
```
Same questions about the lower K limit.

damonge replied:
Fixed in last commit

beckermr commented on src/ccl_power.c:
```diff
@@ -1358,47 +1289,46 @@ static void ccl_cosmology_compute_power_emu(ccl_cosmology * cosmo, int * status)
+	break;
       }
       for (int i=0; i<NK_EMU; i++){
 	logx[i] = log(mode[i]);
```
I cannot find where `mode` is defined. Is this a global var? Does it need a more descriptive name?

damonge replied:
`include/ccl_emu17_params.h`. It is a global variable indeed (which hasn't been added in this PR)

damonge commented:
@beckermr I've made the following modifications to (mostly) ccl_power.c:
- Simplified the code so that BBKS and EH power spectra are now computed with the same method (so we don't have to repeat all the splining code)
- Simplified the emulator pk calculator to avoid recalculating a bunch of things and to avoid rewriting the CLASS splining code.
- Simplified the CLASS interface a bit
- Changed some of the variable names to make them more transparent
- Fixed the bug in the BBKS transfer function noted in #591 

I'll move on to solving all conflicts with the master branch and then to simplifying the C_ell interface to address #553

beckermr commented:
@damonge. Looking great! Let’s address #553 in another PR.

damonge commented:
OK, great @beckermr !
I think I've addressed all of the comments, so happy to have another review whenever you have time

beckermr commented on include/ccl_core.h:
```diff
@@ -11,6 +11,37 @@
+  double growth_factor_0; /**< Constant extrapolating growth factor*/
+  gsl_spline2d *pk; /**< Spline holding the values of P(k,a)*/
+} ccl_p2d_t;
+
```
Whoops. I think we might have gotten confused between various comments. These should be in `ccl_ps2d.h` right? I only meant the ones you moved below.

damonge replied:
so a `ccl_cosmology` contains `ccl_p2d_t` structs, and at the same time the functions that use `ccl_p2d_t` structs also use ccl_cosmology objects. If I put the definition of `ccl_p2d_t` in `ccl_p2d.h` I get a circular dependence between `ccl_core.h` and `ccl_p2d.h`, which is very annoying

beckermr replied:
ah shit. well maybe dump the extra header file or just leave it.

beckermr commented on pyccl/ccl_p2d.i:
```diff
@@ -0,0 +1,52 @@
+				   int *status)
+{
+  ccl_p2d_t *psp=ccl_p2d_t_new(na,aarr,nk,lkarr,pkarr,
+			       1,2,ccl_p2d_cclgrowth,
```
You didn't pass the extrapolation orders here.

damonge replied:
Argh, stupid me. Fixed

beckermr commented on pyccl/cls.py:
```diff
@@ -292,6 +293,8 @@ def angular_cl(cosmo, cltracer1, cltracer2, ell,
         cltracer1, cltracer2 (:obj:`Tracer`): Tracer objects, of any kind.
         ell (float or array_like): Angular wavenumber(s) at which to evaluate
             the angular power spectrum.
+        p_of_k_a (Pk2D object or None): 3D Power spectrum to project. If None,
```
We should use the correct sphinx syntax to link things here.

damonge replied:
Fixed

beckermr commented on pyccl/cls.py:
```diff
@@ -308,27 +311,27 @@ def angular_cl(cosmo, cltracer1, cltracer2, ell,
+    cl, status = lib.angular_cl_vec(
+        cosmo, clt1, clt2, psp, l_limber, l_logstep,
+        l_linstep, ell_use, ell_use.size, status)
+    if ell_use.size==1:
```
We can't test like this since a 1d array will always get coerced to a scalar. You can use `np.isscalar` and test `ell` directly.

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,142 @@
+from .core import check
+import numpy as np
+
+# TODO choices about interpolation/extrapolation
```
This comment can be removed.

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,142 @@
+                 is_logp=True, interp_order_lok=1, interp_order_hik=2, cosmo=None):
+
+        status = 0
+        if(pkfunc is None):  # Initialize power spectrum from 2D array
```
no parens needed here

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,142 @@
+            f, status = lib.p2d_eval_single(self.psp, np.log(k), a, cospass,
+                                            status)
+        else:
+            k_use=np.atleast_1d(k)
```
space around `=` here.

damonge replied:
Fixed

beckermr commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,142 @@
+        """Free memory associated with this Pk2D structure
+        """
+        if hasattr(self, 'has_psp'):
+            if self.has_psp:
```
you might want to check directly on psp argument too? I'm not sure.

damonge replied:
Fixed

beckermr commented on src/ccl_cls.c:
```diff
@@ -451,7 +451,7 @@ static void clt_init_wL(CCL_ClTracer *clt,ccl_cosmology *cosmo,
     }
   }
-
+  
```
i'd be nice to avoid these whitespace changes.

damonge replied:
sorry, my emacs sometimes is a bit stupid. Fixed

beckermr commented on src/ccl_p2d.c:
```diff
@@ -0,0 +1,181 @@
+  if(*status==0) {
+    s2dstatus=gsl_spline2d_init(psp->pk,lk_arr,a_arr,pk_arr,nk,na);
+    if(s2dstatus)
+      *status = CCL_ERROR_SPLINE;
```
If we fail here, we should probably free the memory associated with `psp->pk`.

damonge replied:
mmh, this will be freed up when `psp` is freed up, right?

beckermr replied:
True. Not clear to me the caller will bother if there is an error, but you are correct.

damonge replied:
I don't see how this would lead to a leak (I may be missing something): psp is freed up from python as usual, and this function makes sure that psp->pk is either NULL or has been allocated.

beckermr replied:
yep all true - mostly thinking about usage here. resolving

beckermr commented on src/ccl_power.c:
```diff
@@ -33,17 +70,15 @@ static void ccl_free_class_structs(ccl_cosmology *cosmo,
       *status = CCL_ERROR_CLASS;
-      ccl_cosmology_set_status_message(
-        cosmo, "ccl_power.c: ccl_free_class_structs(): Error freeing CLASS spectra:%s\n", sp->error_message);
+      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_free_class_structs(): Error freeing CLASS spectra:%s\n", sp->error_message);
```
These long lines were changed to be easier to read for a reason. Let's put them back.

damonge replied:
I've broken up several of them.

beckermr commented on src/ccl_power.c:
```diff
@@ -33,17 +70,15 @@ static void ccl_free_class_structs(ccl_cosmology *cosmo,
       *status = CCL_ERROR_CLASS;
-      ccl_cosmology_set_status_message(
-        cosmo, "ccl_power.c: ccl_free_class_structs(): Error freeing CLASS transfer:%s\n", tr->error_message);
+      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_free_class_structs(): Error freeing CLASS transfer:%s\n", tr->error_message);
```
Ditto here.

damonge replied:
Ditto too :-)

beckermr replied:
huh? This line is still rather long.

beckermr commented on src/ccl_power.c:
```diff
@@ -75,8 +110,7 @@ static void ccl_free_class_structs(ccl_cosmology *cosmo,
       *status = CCL_ERROR_CLASS;
-      ccl_cosmology_set_status_message(
-        cosmo, "ccl_power.c: ccl_free_class_structs(): Error freeing CLASS thermo:%s\n", th->error_message);
+      ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_free_class_structs(): Error freeing CLASS thermo:%s\n", th->error_message);
```
Ditto here.

damonge replied:
Done

beckermr commented on src/ccl_power.c:
```diff
@@ -180,56 +214,48 @@ static void ccl_run_class(ccl_cosmology *cosmo,
     *status = CCL_ERROR_CLASS;
-    ccl_cosmology_set_status_message(
-      cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS input:%s\n", errmsg);
+    ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS input:%s\n", errmsg);
```
Ditto here.

damonge replied:
Done

beckermr commented on src/ccl_power.c:
```diff
@@ -180,56 +214,48 @@ static void ccl_run_class(ccl_cosmology *cosmo,
     *status = CCL_ERROR_CLASS;
-    ccl_cosmology_set_status_message(
-      cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS background:%s\n", ba->error_message);
+    ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS background:%s\n", ba->error_message);
```
and here

damonge replied:
Done

beckermr commented on src/ccl_power.c:
```diff
@@ -180,56 +214,48 @@ static void ccl_run_class(ccl_cosmology *cosmo,
     *status = CCL_ERROR_CLASS;
-    ccl_cosmology_set_status_message(
-      cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS thermodynamics:%s\n", th->error_message);
+    ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS thermodynamics:%s\n", th->error_message);
```
and here

damonge replied:
Done

beckermr commented on src/ccl_power.c:
```diff
@@ -180,56 +214,48 @@ static void ccl_run_class(ccl_cosmology *cosmo,
     *status = CCL_ERROR_CLASS;
-    ccl_cosmology_set_status_message(
-      cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS pertubations:%s\n", pt->error_message);
+    ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error running CLASS pertubations:%s\n", pt->error_message);
```
I;m not going to mark the rest. If the code cannot fit into a 13 inch laptop screen in regular sized font on github, then it needs to be broken over multiple lines.

damonge replied:
Done

beckermr commented on src/ccl_power.c:
```diff
@@ -407,277 +431,130 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
 
-  if (*status != CCL_ERROR_CLASS)
+  if (*status==0) {
+    init_parser=1;
```
The way this is setup if we fail to fill the class parameters, the parser will not be freed.

damonge replied:
Argh, sorry. Good catch. This was meant to go in the previous block

beckermr replied:
still needs to be changed. I see a refresh at the top. I will check back in a bit.

damonge replied:
Mmh, I've changed this in the last commit

beckermr commented on src/ccl_power.c:
```diff
@@ -407,277 +431,130 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
+    amax = cosmo->spline_params.A_SPLINE_MAX;
+    na = cosmo->spline_params.A_SPLINE_NA_PK+cosmo->spline_params.A_SPLINE_NLOG_PK-1;
+
+    // The x array is initially k, but will later
```
comment out of date.

damonge replied:
updated

beckermr commented on src/ccl_power.c:
```diff
@@ -13,6 +13,43 @@
+    for(int jj=0;jj<nk;jj++) {
+      double k=exp(lk_arr[jj]);
+      double fbcm=ccl_bcm_model_fka(cosmo,k,a,status);
+      pk2d[jj+nk*ii]+=log(fbcm);
```
So the `correct_bcm` method applies corrections to its pk2d input.

damonge replied:
correct, but we only pass pk2d holding non-linear power spectra to it.

beckermr commented:
flake8 is going to choke on this python. I suggest you run that locally and clean it up

damonge commented:
@beckermr : thanks again and back to you!

damonge commented:
>The drama over flake8 is really not needed. 

Don't mind me, I just enjoy moaning

>One question. Is it possible to mix the EH98 model for the linear power and the CosmicEMU for the non-linear power?

Currently no, I'm afraid.

I think I fixed the leak, but let me know if there are still issues.

beckermr commented:
OK. LGTM!

damonge commented:
Great! I'll wait for travis, and maybe we'll do it live during the telecon later

beckermr commented:
Yup. This is going to make @philbull's PR out of date, yet again. Merging PRs faster and not letting them linger will help with this.

jablazek commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,142 @@
+             sampled (i.e. the resolution of `a_arr` and `lk_arr` is high
+             enough to sample the main features in the power spectrum).
+             For reference, CCL will use bicubic interpolation to evaluate
+             the power spectrum at any intermediate point in k and a.
```
Any choice on interpolation?

damonge replied:
No, bicubic is enforced. We could revisit later on

jablazek commented on pyccl/p2d.py:
```diff
@@ -0,0 +1,142 @@
+        interp_order_lok (int): extrapolation order to be used on k-values
+             below the minimum of the splines (use 0, 1 or 2).
+        interp_order_hik (int): extrapolation order to be used on k-values
+             above the maximum of the splines (use 0, 1 or 2).
```
extrapolation is done fitting a polynomial. Is this in P-k-a space, or logP-logk-loga space? Or is this determined by 'islogP'?

damonge replied:
It's determined by `islogP` precisely to avoid that problem. I'll document better.

jablazek replied:
So if I want log-log interpolation to be done, I have to pass in all log values at the start? This is fine, but should be well documented. (It's the sort of thing I have messed up in the past.)

damonge replied:
Is that not clear enough from the description of `is_logp` ?

jablazek replied:
What is expected is indeed clear. The fact that log interpolation can only be done this way is less clear.

damonge replied:
OK, just clarified that.

damonge commented:
@jablazek : Angular power spectra now take in an arbitrary P(k,a) as a Pk2D object (or use the default non-linear matter power spectrum if you pass `None`). The next step will be to define generic "transfer functions" or "tracers" (haven't decided on the name yet) that can then be used to compute any C_ell you want. That'll be in a separate PR though
